### PR TITLE
Update ngrx store for pinned cards from storage

### PR DESCRIPTION
**Pull request by psybuzz**
Diffbase: https://github.com/tensorflow/tensorboard/pull/4219
Followup: https://github.com/tensorflow/tensorboard/pull/4221

Adds ngrx reducer logic and types to support storing
pinned cards that originate from URL storage. Before
'tagMetadataLoaded' creates real cards, TensorBoard is now
able to keep track of pinned cards from the URL, in the ngrx
store.

A followup will implement the actual deeplink provider.

Googlers, see test sync cl/335717460

### CODE REVIEW

**Comment by psybuzz** [comment]
As per offline discussion, added comments to hopefully explain things, and renamed:
CardInStorage --> CardUniqueInfo
PrePinnedCards --> UnresolvedImportedPinnedCards

Ready for review

File: tensorboard/webapp/metrics/store/metrics_reducers.ts [hunk 1]
```diff
@@ -252,6 +322,40 @@ const {initialState, reducers: routeContextReducer} = createRouteContextedState(
...
+    for (const pinnedCardId of state.pinnedCardToOriginal.keys()) {
+      const {tag, runId, sample} = state.cardMetadataMap[pinnedCardId];
+      seenHashes.add(hashCardUniqueInfo(tag, runId, sample));
+    }
+
+    const hydratedState = partialState as URLDeserializedState;
+    const nextUnresolvedImportedPinnedCards = [] as CardUniqueInfo[];
+    for (const card of [
+      ...state.unresolvedImportedPinnedCards,
+      ...hydratedState.metrics.pinnedCards,
```
Comment (stephanwlee) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:346]: Would you mind leaving a small comment on why we have to read `state.unresolvedImportedPinnedCards`? I now get it that because we use routeContextedState, the unresolved cards can be from previous navigation (e.g., route A -> B -> A)

File: tensorboard/webapp/metrics/store/metrics_reducers_test.ts [hunk 2]
```diff
@@ -1313,4 +1377,103 @@ describe('metrics reducers', () => {
...
+
+    it('does not populate pre-pinned store with duplicates', () => {
+      const beforeState = buildMetricsState({
+        unresolvedImportedPinnedCards: [{tag: 'accuracy'}],
+      });
+      const action = routingActions.stateRehydratedFromUrl({
+        routeKind: RouteKind.EXPERIMENT,
+        partialState: {
+          metrics: {
+            pinnedCards: [{tag: 'accuracy'}],
```
Comment (stephanwlee) [inline tensorboard/webapp/metrics/store/metrics_reducers_test.ts:1449]: also test a case where pinnedCards are just duplicates.

File: tensorboard/webapp/metrics/store/metrics_reducers_test.ts [hunk 3]
```diff
@@ -1313,4 +1377,103 @@ describe('metrics reducers', () => {
       expect(nextState.tagGroupExpanded).toEqual(new Map([['foo', true]]));
     });
   });
+
+  describe('pinned card hydration', () => {
```
Comment (stephanwlee) [inline tensorboard/webapp/metrics/store/metrics_reducers_test.ts:1381]: test a case where run metadata is already populated.

File: tensorboard/webapp/metrics/store/metrics_reducers_test.ts [hunk 4]
```diff
@@ -254,6 +255,69 @@ describe('metrics reducers', () => {
       );
     });
 
+    it('automatically pins cards from pre-pinned storage', () => {
```
Comment (stephanwlee) [inline tensorboard/webapp/metrics/store/metrics_reducers_test.ts:258]: add a case where the unresolvedImportedPinnedCards are inconsistent with the cardMetadata. 

for instance, missing step/runId for image based tag.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts [hunk 5]
```diff
@@ -252,6 +322,40 @@ const {initialState, reducers: routeContextReducer} = createRouteContextedState(
...
+    ]) {
+      const hash = hashCardUniqueInfo(card.tag, card.runId, card.sample);
+      if (!seenHashes.has(hash)) {
+        seenHashes.add(hash);
+        nextUnresolvedImportedPinnedCards.push(card);
+      }
+    }
+    return {
+      ...state,
+      unresolvedImportedPinnedCards: nextUnresolvedImportedPinnedCards,
```
[the remark this answers is not in the document]
Comment (psybuzz) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts]: I agree, added logic to resolve cards, here.

In practice, I don't know if tagMetadata can ever load before initial URL rehydration, but if rehydration is more async later, this will be a good defensive measure.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts (hunk 1 above)
Comment (psybuzz) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:305]: Done

File: tensorboard/webapp/metrics/types.ts [hunk 6]
```diff
@@ -46,8 +46,33 @@ export type NonPinnedCardId = string;
...
+ */
 export type CardId = NonPinnedCardId | PinnedCardId;
 
 export type CardIdWithMetadata = CardMetadata & {
   cardId: CardId;
 };
+
+/**
+ * The most minimal representation of a card that uniquely identifies it.
+ * This information can be used in storage layers (e.g. URL) and may be matched
```
[the remark this answers is not in the document]
Comment (psybuzz) [reply tensorboard/webapp/metrics/types.ts]: > This information may be persisted...

Done

> Also, maybe it is worth noting potential staleness/incorrectness (since user may edit it by hand) of the data

Same question, are we referring to the case of deleting runs only?

File: tensorboard/webapp/metrics/store/metrics_reducers.ts [hunk 7]
```diff
@@ -312,23 +389,36 @@ const reducer = createReducer(
...
         if (!state.cardMetadataMap.hasOwnProperty(cardId)) {
           nextCardMetadataMap[cardId] = cardMetadata;
-          nextCardList.push(cardId);
+          newCardIds.push(cardId);
         }
       }
 
+      const nextCardList = newCardIds.length
+        ? [...state.cardList, ...newCardIds]
+        : state.cardList;
```
Comment (stephanwlee) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:405]: can't you spread zero length array? `const nextCardList = [...state.cardList, ...newCardIds]` should suffice?

File: tensorboard/webapp/metrics/store/metrics_reducers_test.ts [hunk 8]
```diff
@@ -1313,4 +1421,165 @@ describe('metrics reducers', () => {
...
+      });
+      const nextState = reducers(beforeState, action);
+
+      const pinnedCopyId = getPinnedCardId('card1');
+      expect(nextState.pinnedCardToOriginal).toEqual(
+        new Map([[pinnedCopyId, 'card1']])
+      );
+      expect(nextState.cardToPinnedCopy).toEqual(
+        new Map([['card1', pinnedCopyId]])
+      );
```
Comment (stephanwlee) [inline tensorboard/webapp/metrics/store/metrics_reducers_test.ts:1497]: assert that `unresolvedImportedPinnedCards` is empty.

File: tensorboard/webapp/metrics/types.ts [hunk 9]
```diff
@@ -46,8 +46,34 @@ export type NonPinnedCardId = string;
...
+/**
+ * The most minimal representation of a card that uniquely identifies it across
+ * a browser session. This information may be persisted in storage, retrieved,
+ * and used to match against an existing card with the same metadata.
+ */
+export interface CardUniqueInfo {
+  plugin: string;
+  tag: string;
+  runId?: string;
+  sample?: number;
```
Comment (stephanwlee) [inline tensorboard/webapp/metrics/types.ts:69]: this is `maxSample` right? Or is this current sample number?

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts [hunk 10]
```diff
@@ -153,3 +166,125 @@ export function getRunIds(
...
+  return (
+    cardMetadata.plugin === cardUniqueInfo.plugin &&
+    cardMetadata.tag === cardUniqueInfo.tag &&
+    cardMetadata.sample === cardUniqueInfo.sample &&
+    (cardMetadata.runId === cardUniqueInfo.runId || noRunId)
+  );
+}
+
+/**
+ * Attempts to resolve the imported pins against the list of non-pinned cards
```
Comment (stephanwlee) [inline tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts:188]: Not a big comment person but let's codified in the comment that the contract is not to sanitize/clip/validate unresolved cards.

**Review by stephanwlee — APPROVED** [review]

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts (hunk 10 above)
Comment (psybuzz) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts:188]: Done.

**Review by psybuzz — COMMENTED** [review]
Thanks for reviewing!
